### ISSUE TITLE
(feature request) add "download all files" from share/folder

### ISSUE THREAD

**Issue by metalheim** [proofrock/pupcloud#34]
tldr: bulk-downloads of all files in a folder/share

## My User Story
I have a folder full of pictures that i want to share with a friend. (f.e. the photos from a shared vacation)

- [x] I want my friend to be able to view the folder and individidual files in the browser (that works great already)
- [ ] I want my friend to be able to download all files from the folder. Currently, he would need to download each image individually.

## How to implement & considerations
I see two options really:
1. All files are concetenated on-the-fly into 1 big file (f.e. a .zip or .tar file)
2. All files are downloaded individually 

Option 1. is the nicer one imho, its also how most other services do it (f.e. Google drive, Dropbox etc.). But it might need a cache directory? That would kinda contradict the minimalist "leave no marks on your filesystem" goal of pupcloud.

Option 2. has no dependencies, but browsers do not like a client-side invocation of lots of file downloads. (Because Its not very userfriendly - what happens on name conflicts etc.)

**Comment by proofrock** [comment]
Thanks, and thanks for the very structured report (do you have a background in UX?). I'd like to implement the first option, a cache is not necessary because what I want to do is to feed a zip stream into the server's network output... it's very easy to do in Java, I don't know in Go, finger crossed ;-)

**Comment by metalheim** [comment]
I dont have a background in UX, or programming really. (unfortunately, my company would love that ;) )
I'm just trying to contribute with well-writen tickets at least.

Another nice OS software written in GO is "navidrome", which also has a bundle-download zip functionality. Maybe you can get some inspiration how it does it:

https://github.com/navidrome/navidrome/blob/31c598de075fc7af96be93070056a6666da1b5c3/core/archiver.go
https://github.com/navidrome/navidrome/blob/8afa2cd833c4fdf1a72a4debb1570281434e2af0/server/subsonic/stream.go#L82

**Comment by proofrock** [comment]
Note for my future self:

```go
func (a *archiver) zipTracks(ctx context.Context, id string, out io.Writer, mfs model.MediaFiles, ch createHeader) error {
	z := zip.NewWriter(out)
	for idx, mf := range mfs {
		_ = a.addFileToZip(ctx, z, mf, ch(idx, mf))
	}
	err := z.Close()
	if err != nil {
		log.Error(ctx, "Error closing zip file", "id", id, err)
	}
	return err
}
```

Thanks, now I know how :-)

**Comment by proofrock** [comment]
Our UX designer always bashes us on the head with user stories, that's why I was asking 😉 Anyway, thanks for a great report. I should write a sort of guide, but really... I don't want to impose too many constraints on people, and everyone is amazingly well-behaved (and I am used to a bit of chaos myself). Still, it's nice to see.

**Comment by proofrock** [comment]
Closing, as this was moved to the [roadmap](https://github.com/proofrock/pupcloud/wiki/Roadmap!). Thanks!
